src/modules/formBuilder/components/formBuilderUtil.ts

```diff
@@ -90,6 +90,137 @@ export const updateFormItem = (
   return copy;
 };
 
+function moveIntoGroup(
+  itemToMove: FormItem,
+  group: FormItem,
+  addMethod: 'push' | 'unshift',
+  thisLayer: FormItem[],
+  currentIndex: number
+) {
+  if (!group.item) group.item = [];
+  group.item[addMethod](itemToMove);
+  thisLayer.splice(currentIndex, 1);
+  return true;
+}
+
+function swapItems(index1: number, index2: number, thisLayer: FormItem[]) {
+  [thisLayer[index1], thisLayer[index2]] = [
+    thisLayer[index2],
+    thisLayer[index1],
+  ];
+  return true;
+}
+
+function moveToParent(
+  itemToMove: FormItem,
+  parentLayer: FormItem[],
+  newIndex: number,
+  thisLayer: FormItem[],
+  currentIndex: number
+) {
+  parentLayer.splice(newIndex, 0, itemToMove);
+  thisLayer.splice(currentIndex, 1);
+  return true;
+}
+
+export const reorderFormItems = (
+  formDefinition: FormDefinitionJson,
+  itemToMove: FormItem,
+  direction: 'up' | 'down'
+) => {
+  const copy = cloneDeep(formDefinition);
+  const toExpand: string[] = [];
+
+  function recurseFindAndMove(
+    thisLayer: Maybe<FormItem[]> | undefined,
+    parentLayer: Maybe<FormItem[]> | undefined,
+    indexInParent: number,
+    nestingDepth: number
+  ): boolean {
+    if (!thisLayer) return false;
+
+    const currentIndex = thisLayer.findIndex(
+      (i) => i.linkId === itemToMove.linkId
+    );
+
+    if (currentIndex >= 0) {
+      // Found the item in this layer
+      if (direction === 'up') {
+        if (currentIndex > 0) {
+          // This item isn't the first item in the list
+          const prevItem = thisLayer[currentIndex - 1];
+
+          if (prevItem.type === ItemType.Group && nestingDepth < 5) {
+            toExpand.push(prevItem.linkId);
+            // If the previous item is a group, and we haven't reached max nesting depth, then move it into the group
```

Comment:
Phew I think I need an example to fully grasp this!

In the below example, assume all items are "questions" except the ones that have nested items, which are groups.
```
{
  "item": {
    { "link_id": 1 },
    { "link_id": 2 },
    { "link_id": 3,
      "item": [
        { "link_id": 4 },
        { "link_id": 5 },
        { "link_id": 6 },
      ]
    },
    { "link_id": 7,
      "item": [
        { "link_id": 8 },
        { "link_id": 9 },
        { "link_id": 10 },
      ]
    },
    { "link_id": 11}
  }
}
```
Are my below assumptions correct? This is not an order of events, each is a separate operation on its own.

* Moving item 3 "up" will swap items 2 and 3
* Moving item 4 "up" will insert item 4 after item 2
* Moving item 7 "up" will insert item 7 after item 6 (in other words, "append" to item 3)
* Moving item 8 "up" will insert item 8 after item 3 ("pops out" of group, moves down a nesting level)
* Moving item 11 "up" will insert item 11 after item 10 (in other words, "append" to item 7) (moves up a nesting level)


I think that lets you move an item to any other position, but it is laborious to move something up far. E.g. if you want to move item 11 to the top, you need to step through all the nested items in all the groups above it.

No changes requested, this is purely for validating my understanding. Thanks!

Reply:
Yes, I think what you've described here is what I've tried to implement, basically trying to match the prototype Bobby created here: (see link, but I'm linking his comment for background) https://github.com/open-path/Green-River/issues/5793#issuecomment-2112462489

I agree it is super laborious for items to move far. Keyboard shortcuts would be a big improvement, drag/drop would be even better.

Reply:
Okay gotcha, thanks 👍